NaN in data cause json serialization to fatal

fortio version: docker package->istio/fortio:latest from 01/23/2018

18:11:12 F fortio_main.go:273> Unable to json serialize result: json: unsupported value: NaN
panic: aborting...
goroutine 1 [running]:
istio.io/fortio/log.logPrintf(0x6, 0x931b21, 0x23, 0xc420205a50, 0x1, 0x1)
	/go/src/istio.io/fortio/log/logger.go:160 +0x28c
istio.io/fortio/log.Fatalf(0x931b21, 0x23, 0xc420205a50, 0x1, 0x1)
	/go/src/istio.io/fortio/log/logger.go:208 +0x5c
main.fortioLoad()
	/go/src/istio.io/fortio/fortio_main.go:273 +0xb5c
main.main()
	/go/src/istio.io/fortio/fortio_main.go:144 +0x669

--- a/stats/stats.go
+++ b/stats/stats.go
@@ -68,6 +68,11 @@ func (c *Counter) Avg() float64 {
 func (c *Counter) StdDev() float64 {
 	fC := float64(c.Count)
 	sigma := (c.sumOfSquares - c.Sum*c.Sum/fC) / fC
+	// should never happen but it does
+	if sigma < 0 {
+		log.Warnf("Unexpected negative sigma for %+v: %g", c, sigma)
+		return 0
+	}
 	return math.Sqrt(sigma)
 }
 

--- a/stats/stats_test.go
+++ b/stats/stats_test.go
@@ -708,6 +708,18 @@ func TestRound(t *testing.T) {
 	}
 }
 
+func TestNaN(t *testing.T) {
+	var c Counter
+	for i := 599713; i > 0; i-- {
+		c.Record(1281)
+	}
+	c.Log("counter with 599713 times 1281 - issue #97")
+	actual := c.StdDev()
+	if actual != 0.0 {
+		t.Errorf("Got %g, expected 0 for stddev/issue #97 c is %+v", actual, c)
+	}
+}
+
 // TODO: add test with data 1.0 1.0001 1.999 2.0 2.5
 // should get 3 buckets 0-1 with count 1
 // 1-2 with count 3

